engine/equivalence.ml

```diff
@@ -0,0 +1,85 @@
+open Merge_accessors
+
+type specialized_tree =
+  | Terminal of constraint_tree
+  | SNode of sym_value * (node_constraint * specialized_tree) list * (node_constraint * specialized_tree) option
+and
+  node_constraint =
+  | Symbolic of domain
+  | Resolved of bool
+
+let compare (src: Merge_accessors.constraint_tree) (dst: Merge_accessors.constraint_tree) : bool =
+  let rec init : Merge_accessors.constraint_tree -> specialized_tree = function
+    | (Leaf _ | Failure) as t -> Terminal t
+    | Node (acc, children, fallback) ->
+      let c = children |> List.map (fun (pi, c_tree) -> Symbolic pi, init c_tree) in
+      let fb = fallback |> Option.map (fun (pi, c_tree) -> Symbolic pi, init c_tree) in
+      SNode (acc, c, fb)
+  in
+  let specialize (pi: domain) (pi': domain) : node_constraint =
+    let intersect = Target_sym_engine.Domain.inter in
+    let is_empty (set:domain) =
+      Target_sym_engine.Domain.Set.is_empty set.int && Target_sym_engine.Domain.Set.is_empty set.tag
+    in
+    match intersect pi pi' with
+    | x when x = pi' ->
+      Resolved true
+    | x when is_empty x ->
+      Resolved false
+    | x -> Symbolic x
+  in
+  let rec _compare (src: Merge_accessors.constraint_tree) (dst: specialized_tree) : bool =
+    let rec trim src_acc src_constraints = function
+      | SNode (_, [], None) -> assert false
+      | Terminal c -> Terminal c
+      | SNode (acc, children, fallback) ->
+        let children' =
+          children
+          |> List.filter_map (fun (pi, s_tree) ->
+              match pi with
+              | Resolved false -> None
+              | Resolved true -> Some (pi, trim src_acc src_constraints s_tree)
+              | Symbolic dst_constraint ->
+                if src_acc = acc then
+                  Some (specialize src_constraints dst_constraint, trim src_acc src_constraints s_tree)
+                else
+                  Some (pi, trim src_acc src_constraints s_tree))
```

Comment:
Trimming is incomplete here: you correctly remove all children whose condition was resolved to `false`, but you keep the children that resolve to `true` with the others. What we discussed is that, when one child resolves to `true`, you can drop all other children and the node structure and just return this child's subtree. (Similarly, when the fallback becomes true, you can drop the node structure and return the fallback's subtree.)

Maybe one way to do this would be to first use a `filter_map` to remove the false nodes as you are doing (and filter out the fallback if it resolves to true), and then use a `BatList.find_map` to find the true nodes (and look also at the fallback), and if that fails rebuild the node structure. (Note: if we don't have `Resolved of bool` anymore, we can directly test for the inclusion of the source/dividing/assumed domain within the local/child domain.)

Currently you seem to be dealing with this at the `compare_terminals` level, but I don't think that your code there is correct -- it correctly deals with `Resolved true` if it appears in the first child of a node, but what about cases where the resolution is only in a later child? It could still be correct due to the fact that node conditions are non-overlapping, but this would be needlessly confusing.